### PR TITLE
Improved UnsupporedProtocol error messaging

### PR DESCRIPTION
**Pull request by default-303**
fixes #1644 in [https://github.com/encode/httpx/issues/1644](encode/httpx)

### CODE REVIEW

**Comment by tomchristie** [comment]
Okay, so you'll need to run `scripts/lint` to fix the listing issues here.

(That'll also autogenerate the sync versions of the code)

**Comment by default-303** [comment]
okay, ive run the scripts. Should work now

**Comment by default-303** [comment]
I checked the details turns out its cuz the line was too long, so i naturally escaped the string with `/` but running `scripts/test` is now giving this error -
```
Success: no issues found in 51 source files
+ scripts/unasync --check
httpcore/_async/connection_pool.py -> httpcore/_sync/connection_pool.py
unasync mismatch between 'httpcore/_async/connection_pool.py' and 'httpcore/_sync/connection_pool.py'
Async code:         '                    f"The request to \'://{host}/\' is missing either an \'http://\' \\\n'
Expected sync code: '                    f"The request to \'://{host}/\' is missing either an \'http://\' \\\n'
Actual sync code:   '                    f"The request to \'://{host}/\' is missing either an \'http://\' or \'https://\' protocol."\n'
```
What to do now ?

File: httpcore/_async/connection_pool.py [hunk 1]
```diff
@@ -201,7 +201,16 @@ async def handle_async_request(
     ) -> Tuple[int, Headers, AsyncByteStream, dict]:
         if url[0] not in (b"http", b"https"):
             scheme = url[0].decode("latin-1")
-            raise UnsupportedProtocol(f"Unsupported URL protocol {scheme!r}")
+            host = url[1].decode("latin-1")
+            if scheme == "":
+                raise UnsupportedProtocol(
+                    f"The request to '://{host}/' is missing either an 'http://' or 'https://' protocol."
```
Comment (tomchristie) [inline httpcore/_async/connection_pool.py:207]: Looks like this string [needs to be split across two lines](https://github.com/encode/httpcore/pull/354/checks?check_run_id=2781908584).

File: httpcore/_async/connection_pool.py [hunk 2]
```diff
@@ -201,7 +201,16 @@ async def handle_async_request(
...
+                )
+            else:
+                raise UnsupportedProtocol(
+                    f"The request to '{scheme}://{host}' has an unsupported protocol {scheme!r}"
```
Comment (tomchristie) [inline httpcore/_async/connection_pool.py:211]: Same here.

**Comment by default-303** [comment]
okay, thanks. running `scripts/lint` gets rid of the above error.

**Comment by default-303** [comment]
hey, can you approve the tests again. This should do it 🤞

**Comment by default-303** [comment]
are the changes good ? let me know if I need to tweak anything 😀

File: tests/sync_tests/test_http2.py [hunk 3]
```diff
@@ -247,164 +247,3 @@ def test_request_with_missing_host_header() -> None:
...
-#         # read returns `b""`, indicating a premature close.
-#         with pytest.raises(httpcore.RemoteProtocolError) as excinfo:
-#             http.handle_request(
-#                 method=b"GET",
-#                 url=(b"http", b"example.org", None, b"/"),
-#                 headers=[(b"Host", b"example.org")],
-#                 stream=httpcore.ByteStream(b""),
-#                 extensions={},
-#             )
-#         assert str(excinfo.value) == "Server disconnected without sending a response."
```
Comment (tomchristie) [inline tests/sync_tests/test_http2.py:410]: Not sure how this leaked in, but it's my fault it's there.
We might sometime deal with this unrelated cleanup in a separately targeted PR, but let's just roll with it here.
Keep things simple, keep things moving.

**Comment by tomchristie** [comment]
Great, thank you!

File: tests/sync_tests/test_http2.py (hunk 3 above)
Comment (default-303) [reply tests/sync_tests/test_http2.py:410]: sure, raise a new issue itll be cool to help

**Pull request merged**